x/emissions/keeper/topic_activation.go

```diff
@@ -0,0 +1,346 @@
+		// Remove the topic with the lowest weight
+		lowestWeight, _, err := k.GetLowestActiveTopicWeightAtBlock(ctx, block)
+		if err != nil {
+			return false, err
+		}
+
+		weight, err := k.GetTopicWeightFromTopicId(ctx, topicId)
+		if err != nil {
+			return false, err
+		}
+
+		if weight.Lt(lowestWeight.Weight) {
+			sdkCtx.Logger().Warn(fmt.Sprintf("Topic%d cannot be activated due to less than lowest weight at block %d", topicId, block))
+			return false, nil
+		}
+		_, err = k.inactivateTopicWithoutMinWeightReset(ctx, lowestWeight.TopicId)
+		if err != nil {
+			return false, err
+		}
+
+		// Remove the lowest weight topic from the active topics at the block
+		for i, id := range existingActiveTopics {
+			if id == lowestWeight.TopicId {
+				existingActiveTopics = append(existingActiveTopics[:i], existingActiveTopics[i+1:]...)
+				break
+			}
+		}
+	}
+
+	existingActiveTopics = append(existingActiveTopics, topicId)
+	// Add newly active topic to the active topics at the block
+	newActiveTopicIds := types.TopicIds{TopicIds: existingActiveTopics}
+	err = k.blockToActiveTopics.Set(ctx, block, newActiveTopicIds)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// Set a topic to active if the topic exists, else does nothing
+func (k *Keeper) ActivateTopic(ctx context.Context, topicId TopicId) error {
+	topicExists, err := k.topics.Has(ctx, topicId)
+	if err != nil {
+		return err
+	}
+	if !topicExists {
+		return nil
+	}
+
+	// Check topic activation with next possible churning block
+	_, topicIsActive, err := k.GetNextPossibleChurningBlockByTopicId(ctx, topicId)
+	if err != nil {
+		return err
+	}
+	if topicIsActive {
+		return nil
+	}
+
+	topic, err := k.GetTopic(ctx, topicId)
+	if err != nil {
+		return err
+	}
+	currentBlock := sdk.UnwrapSDKContext(ctx).BlockHeight()
+	epochEndBlock := currentBlock + topic.EpochLength
+
+	// Add this topic with epochend block
+	// If the topic of the epochend block exceeds the limit, remove topic with lowest weight
+	isAdded, err := k.addTopicToActiveSetRespectingLimitsWithoutMinWeightReset(ctx, topicId, epochEndBlock)
+	if err != nil {
+		return err
+	}
+	if !isAdded {
+		return nil
+	}
+
+	err = k.topicToNextPossibleChurningBlock.Set(ctx, topicId, epochEndBlock)
+	if err != nil {
+		return err
+	}
+
+	// Update lowest topic weight of the block
+	err = k.ResetLowestActiveTopicWeightAtBlock(ctx, epochEndBlock)
+	if err != nil {
+		return err
+	}
+
+	// Set active for this topic
+	err = k.activeTopics.Set(ctx, topicId)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Inactivate the topic
+func (k *Keeper) InactivateTopic(ctx context.Context, topicId TopicId) error {
+	_, err := k.inactivateTopicWithoutMinWeightReset(ctx, topicId)
+	return err
+}
+
+// If the topic weight is not less than lowest weight keep it as activated
+func (k *Keeper) AttemptTopicReactivation(ctx context.Context, topicId TopicId) error {
```

Comment:
most of this function is the same of the `ActivateTopic`, prob it can be refactored